book/src/design/implementation/selector-combining.md

```diff
@@ -0,0 +1,116 @@
+   column.
+2. Check (or ensure by construction) that all polynomial constraints involving each simple
+   selector $s_i$ are of the form $s_i \cdot t_{i,j} = 0$ where $t_{i,j}$ do not involve
+   any simple selectors. For each $i$, record the maximum degree of any $t_{i,j}$ as
+   $d^\mathsf{max}_i$.
+3. Compute a binary "exclusion matrix" $X$ such that $X_{j,i}$ is $1$ whenever $i \neq j$
+   and $s_i$ and $s_j$ are enabled on the same row; and $0$ otherwise.
+   > Since $X$ is symmetric and is zero on the diagonal, we can represent it by either its
+   > upper or lower triangular entries. The rest of the algorithm is guaranteed only to
+   > access only the entries $X_{j,i}$ where $j > i$.
+4. Initialize a boolean array $\mathsf{added}_{0..{k-1}}$ to all $\mathsf{false}$.
+   > $\mathsf{added}_i$ will record whether $s_i$ has been included in any combination.
+6. Iterate over the $s_i$ that have not yet been added to any combination:
+   * a. Add $s_i$ to a fresh combination $c$, and set $\mathsf{added}_i = \mathsf{true}$.
+   * b. Let mut $d := d^\mathsf{max}_i - 1$.
+     > $d$ is used to keep track of the largest degree, *excluding* the selector
+     > expression, of any gate involved in the combination $c$ so far.
+   * c. Iterate over all the selectors $s_j$ for $j > i$ that can potentially join $c$,
+     i.e. for which $\mathsf{added}_j$ is false:
+     * i. (Optimization) If $d + \mathsf{len}(c) = \mathsf{max\_degree}$, break to the
+       outer loop, since no more selectors can be added to $c$.
+     * ii. Let $d^\mathsf{new} = \mathsf{max}(d, d^\mathsf{max}_j-1)$.
+     * iii. If $X_{j,i'}$ is $\mathsf{true}$ for any $i'$ in $c$, or if
+       $d^\mathsf{new} + (\mathsf{len}(c) + 1) > \mathsf{max\_degree}$, break to the outer
+       loop.
+       > $d^\mathsf{new} + (\mathsf{len}(c) + 1)$ is the maximum degree, *including* the
+       > selector expression, of any constraint that would result from adding $s_j$ to the
+       > combination $c$.
+     * iv. Set $d := d^\mathsf{new}$.
+     * v. Add $s_j$ to $c$ and set $\mathsf{added}_j := \mathsf{true}$.
+   * d. Allocate a fixed column $q_c$, initialized to all-zeroes.
+   * e. For each selector $s' \in c$:
+     * i. Label $s'$ with a distinct index $k$ where $1 \leq k \leq \mathsf{len}(c)$.
+     * ii. Record that $s'$ should be substituted with
+       $q_c \cdot \prod_{1 \leq h \leq \mathsf{len}(c),\,h \neq k} (h-q_c)$ in all gate
+       constraints.
+     * iii. For each row $r$ such that $s'$ is enabled at $r$, assign the value $k$ to
+       $q_c$ at row $r$.
+
+The above algorithm is implemented in
+[halo2_proofs/src/plonk/circuit/compress_selectors.rs](https://github.com/zcash/halo2/blob/main/halo2_proofs/src/plonk/circuit/compress_selectors.rs).
+This is used by the `compress_selectors` function of
+[halo2_proofs/src/plonk/circuit.rs](https://github.com/zcash/halo2/blob/main/halo2_proofs/src/plonk/circuit.rs)
+which does the actual substitutions.
```

Comment:
We'll want to link to files by their revision so these links don't rot.

Reply:
Hmm, I intentionally linked to the whole file so that they would be less likely to rot. I think revision links are less useful here. If we're doing a major refactoring of the selector combining code then we would probably want to update this doc anyway.